### PR TITLE
Only require step code if requirements exist

### PR DESCRIPTION
**Pull request by LaurelOlson**
## Description

Refactor form json into a separate service and add in functionality to transform the step code so it is required or not depending on the presence of step requirements.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📦 Chore (Release)
- [ ] ✅ Test

## Related Tickets & Documents

[BPHH-1489](https://hous-bssb.atlassian.net/browse/BPHH-1489)

## Steps to QA

Login as a review manager and update step requirements for a permit type so both are "not required". This should remove the validation on the step code field in the permit application form. Setting either energy or zero carbon requirements back to a value should make the step code required again. 

## UI Accessibility Checklist

_Note the following checklist is not an exhaustive list. Check
the [WebAIM checklist](https://webaim.org/standards/wcag/checklist) for
a user-friendly checklist with more details. Aim to meet at least *AA* conformance level where applicable. Check
the [WCAG guidelines](https://webaim.org/standards/wcag/checklist) for the full guidelines._

- [ ] Markup uses semantic HTML?
- [ ] Additional context added through `aria-roles` and `aria-labels`?
- [ ] Checked
      with [Wave Browser Extension](https://chromewebstore.google.com/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh)
      for accessibility errors?
- [ ] Usable with a keyboard and navigable in a logical order?
- [ ] Usable using a screen reader (e.g. Voiceover for Mac) with enough context for the user?

## General Checklist

- [ ] ✅ Provide tests for your changes where relevant
- [x] 📝 Use descriptive commit messages
- [ ] 📗 Update any related documentation and include any relevant screenshots
- [ ] 📱 For UI-related tasks, ensure responsive design is implemented across multiple screen size
- [x] 🗂️ Move your relevant story/task to the _Ready for Code Review_ state

## [optional] Are there any post-deployment tasks we need to perform?

n/a

### CODE REVIEW

File: app/models/permit_application.rb [hunk 1]
```diff
@@ -217,6 +207,14 @@ def missing_pdfs
     missing_pdfs
   end
 
+  def energy_step_code_required?
+    jurisdiction
+      .permit_type_required_steps
+      .where(permit_type_id:)
+      .customizations
+      .all? { |rs| !rs.energy_step_required && !rs.zero_carbon_step_required }
```
Comment (taw27) [inline app/models/permit_application.rb:215]: so all of energy_step_required and zero_carbon_step_required have to be false for energy step_code to be required?

**Review by taw27 — APPROVED** [review]

File: app/models/permit_application.rb (hunk 1 above)
Comment (LaurelOlson) [reply app/models/permit_application.rb:215]: The real world expectation is that if it's not required, there will only be one customization and both values will be nil. It wouldn't make sense if there are multiple required steps where one of them has both set to not required but if that is the case, we would treat it as required as there is a set of requirements.

Comment (LaurelOlson) [reply app/models/permit_application.rb:215]: OK actually yes this logic is flawed, will update

Comment (LaurelOlson) [reply app/models/permit_application.rb]: @taw27 OK fixed, should be easier to make sense of now and works correctly

Comment (taw27) [reply app/models/permit_application.rb]: @LaurelOlson what happens if only one  of `energy_step_required ` or `zero_carbon_step_required ` is nil? This  would be energy_step_code_required? is false?

**Review by taw27 — APPROVED** [review]

File: app/models/permit_application.rb (hunk 1 above)
Comment (taw27) [reply app/models/permit_application.rb]: @LaurelOlson approved after latest change

**Pull request merged**